packages/core/src/modules/message-pickup/protocol/v1/V1MessagePickupProtocol.ts

```diff
@@ -127,4 +133,40 @@ export class V1MessagePickupProtocol extends BaseMessagePickupProtocol {
 
     return new OutboundMessageContext(batchMessage, { agentContext: messageContext.agentContext, connection })
   }
+
+  public async processBatch(messageContext: InboundMessageContext<V1BatchMessage>) {
+    const { message: batchMessage, agentContext } = messageContext
+    const { messages } = batchMessage
+
+    const connection = messageContext.assertReadyConnection()
+
+    const eventEmitter = messageContext.agentContext.dependencyManager.resolve(EventEmitter)
+
+    messages.forEach((message) => {
+      eventEmitter.emit<AgentMessageReceivedEvent>(messageContext.agentContext, {
+        type: AgentEventTypes.AgentMessageReceived,
+        payload: {
+          message: message.message,
+          contextCorrelationId: messageContext.agentContext.contextCorrelationId,
+        },
+      })
+    })
+
+    // A Batch message without messages at all means that we are done with the
+    // message pickup process (Note: this is not optimal since we'll always doing an extra
+    // Batch Pickup. However, it is safer and should be faster than waiting an entire loop
+    // interval to retrieve more messages)
+    if (messages.length === 0) {
```

Comment:
If messages.length is smaller than the batch size we can also assume the message pickup is complete right?

Reply:
Yes! The issue is that I cannot be 100% sure of the batch size requested in the originating batch message (I can assume it from agent config but user can actually override it)